x/contracts/wasmlanche/wasmlanche-test/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+use borsh::BorshSerialize;
+use std::{
+    cell::RefCell,
+    collections::{HashMap, HashSet},
+    path::PathBuf,
+    process::Command,
+};
+use wasmlanche::{Address, ID_LEN};
+use wasmtime::{
+    Caller, Config, Engine, Extern, Instance, Linker, Memory, Module, OptLevel, Store, StoreLimits,
+    StoreLimitsBuilder, TypedFunc,
+};
+
+const WASM_TARGET: &str = "wasm32-unknown-unknown";
+const PROFILE: &str = "release";
+const ALLOC_FN_NAME: &str = "alloc";
+
+type AllocParam = u32;
+type AllocReturn = u32;
+type AllocFn = TypedFunc<AllocParam, AllocReturn>;
+pub type UserDefinedFnParam = u32;
+pub type UserDefinedFnReturn = ();
+pub type UserDefinedFn = TypedFunc<UserDefinedFnParam, UserDefinedFnReturn>;
+type StateKey = Box<[u8]>;
+type StateValue = Box<[u8]>;
+
+pub struct StoreData {
+    call_result: Option<Vec<u8>>,
+    state: HashMap<StateKey, StateValue>,
+    limiter: StoreLimits,
+}
+
+impl StoreData {
+    #[inline]
+    fn set_result(&mut self, result: Vec<u8>) {
+        self.call_result = Some(result);
+    }
+
+    #[inline]
+    fn set<Iter: IntoIterator<Item = (StateKey, StateValue)>>(&mut self, iter: Iter) {
+        self.state.extend(iter);
+    }
+
+    #[inline]
+    fn get(&self, key: &[u8]) -> Option<&StateValue> {
+        self.state.get(key)
+    }
+
+    #[inline]
+    pub fn take_result(&mut self) -> Option<Vec<u8>> {
+        self.call_result.take()
+    }
+}
+
+pub struct Builder {
```

Comment:
something feels strange to me regarding the relationship of `Builder` and `TestCrate`. Seems like `TestCrate::new(crate_name)` and `TestCrate::build(&self)` would make more sense. We wouldn't need the use the `builder` struct that could all be done internally.  

Then 
```rust
 b.iter_batched(
            || TestCrate::new("test-crate"),
            |crate| {
                let mut contract = Contract::new(crate); // new would call crate.build
                contract.always_true();
                contract
            },
            BatchSize::PerIteration,
        )
```

Reply:
The thing that's strange is that I'm cheating the pattern a little. This should really be two builders and a resultant type. 
```rust
// really a module, engine, and linker, but the engine and linker are singletons for us
let module = ModuleBuilder::new("test-crate").build();
let test_instance = module.instantiate(); 
```
Happy to change it to this if you think that would be clearer.

With the naming from your example above, the `TestCrate` is the type that's doing the work to _construct_ the `Contract`, so it's actually a bit of an anti-pattern to have to pass it into a constructor instead of just having a method called `TestCrate::instantiate_contract` or whatever you want to call it here. 

But in the end, you're right, there was definitely a code-smell here.